quarkus:dev cannot pick newly added config property

If you add new property in microprofile-config.properties in quarkus:dev mode this property is not picked in hot reload.

Reproducer: https://github.com/xstefank/learning-tests/tree/master/quarkus/config-new-prop

Here is what I did:

Use:

```
@Path("/greeting")
public class GreetingResource {

    @Inject
    @ConfigProperty(name = "testProp")
    private String testProp;

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return testProp;
    }
}
```
and 

```
testProp = asdf
```

Execute:

```
mvn clean compile
mvn quarkus:dev
```

`http://localhost:8080/greeting` correctly returns: `asdf`

Add `testProp2 = hello` to `microprofile-config.properties`.

change `@ConfigProperty(name = "testProp")` to `@ConfigProperty(name = "testProp2")`.
When doing that I get:

```
Caused by: java.lang.IllegalStateException: No config value exists for: testProp2
```

```
2019-03-18 11:55:38,569 INFO  [io.qua.dev] (XNIO-1 task-1) Changes source files detected, recompiling [/home/gandrian/Desktop/newprop/src/main/java/org/acme/config/GreetingResource.java]
2019-03-18 11:55:38,873 INFO  [io.quarkus] (XNIO-1 task-1) Quarkus stopped in 0.001s
2019-03-18 11:55:38,874 INFO  [io.qua.dep.QuarkusAugmentor] (XNIO-1 task-1) Beginning quarkus augmentation
2019-03-18 11:55:38,926 ERROR [io.qua.dev.DevModeMain] (XNIO-1 task-1) Failed to start quarkus: java.lang.RuntimeException: org.jboss.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: javax.enterprise.inject.spi.DeploymentException: java.lang.IllegalStateException: No config value exists for: testProp2
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:134)
        at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:121)
        at io.quarkus.dev.DevModeMain.restartApp(DevModeMain.java:148)
        at io.quarkus.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:87)
        at io.quarkus.undertow.deployment.devmode.UndertowHotReplacementSetup.handleHotDeploymentRequest(UndertowHotReplacementSetup.java:54)
        at io.quarkus.undertow.deployment.devmode.UndertowHotReplacementSetup$1$1.handleRequest(UndertowHotReplacementSetup.java:35)
        at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
        at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1416)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.jboss.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.arc.deployment.ArcAnnotationProcessor#build threw an exception: javax.enterprise.inject.spi.DeploymentException: java.lang.IllegalStateException: No config value exists for: testProp2
        at org.jboss.builder.Execution.run(Execution.java:123)
        at org.jboss.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:136)
        at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:102)
        at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:99)
        ... 12 more
Caused by: javax.enterprise.inject.spi.DeploymentException: java.lang.IllegalStateException: No config value exists for: testProp2
        at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:624)
        at io.quarkus.arc.processor.BeanDeployment.validate(BeanDeployment.java:256)
        at io.quarkus.arc.processor.BeanProcessor.process(BeanProcessor.java:160)
        at io.quarkus.arc.deployment.ArcAnnotationProcessor.build(ArcAnnotationProcessor.java:243)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:435)
        at org.jboss.builder.BuildContext.run(BuildContext.java:413)
        at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
        at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1416)
        at java.lang.Thread.run(Thread.java:748)
        at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.lang.IllegalStateException: No config value exists for: testProp2
        at io.quarkus.arc.deployment.ConfigBuildStep$ConfigBeanDeploymentValidator.validate(ConfigBuildStep.java:97)
        at io.quarkus.arc.processor.BeanDeployment.validate(BeanDeployment.java:253)
        ... 14 more

2019-03-18 11:55:38,927 INFO  [io.qua.dev] (XNIO-1 task-1) Hot replace total time: 0.360s 

```

Still the same issue for me @mkouba.

```
java.lang.RuntimeException: Failed to start quarkus
	at io.quarkus.runner.ApplicationImpl3.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:93)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:119)
	at io.quarkus.dev.DevModeMain.doStart(DevModeMain.java:105)
	at io.quarkus.dev.DevModeMain.restartApp(DevModeMain.java:132)
	at io.quarkus.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:87)
	at io.quarkus.undertow.deployment.devmode.UndertowHotReplacementSetup.handleHotDeploymentRequest(UndertowHotReplacementSetup.java:56)
	at io.quarkus.undertow.deployment.devmode.UndertowHotReplacementSetup$1$1.handleRequest(UndertowHotReplacementSetup.java:37)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:364)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1998)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1525)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1416)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.enterprise.inject.spi.DeploymentException: No config value of type class java.lang.String exists for: greeting.other
	at io.quarkus.arc.runtime.ConfigDeploymentTemplate.validateConfigProperties(ConfigDeploymentTemplate.java:45)
	at io.quarkus.deployment.steps.ConfigBuildStep$validateConfigProperties26.deploy(Unknown Source)
	... 15 more
```

--- a/core/devmode/src/main/java/io/quarkus/dev/DevModeMain.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/DevModeMain.java
@@ -22,13 +22,11 @@ import java.io.IOException;
 import java.net.URL;
 import java.net.URLClassLoader;
 
-import org.eclipse.microprofile.config.Config;
 import org.jboss.logging.Logger;
 
 import io.quarkus.runner.RuntimeRunner;
 import io.quarkus.runtime.LaunchMode;
 import io.quarkus.runtime.Timing;
-import io.smallrye.config.PropertiesConfigSource;
 import io.smallrye.config.SmallRyeConfigProviderResolver;
 
 /**
@@ -59,7 +57,7 @@ public class DevModeMain {
 
         runtimeUpdatesProcessor = RuntimeCompilationSetup.setup();
         if (runtimeUpdatesProcessor != null) {
-            runtimeUpdatesProcessor.scanForChangedClasses();
+            runtimeUpdatesProcessor.checkForChangedClasses();
         }
         //TODO: we can't handle an exception on startup with hot replacement, as Undertow might not have started
 

--- a/core/devmode/src/main/java/io/quarkus/dev/RuntimeUpdatesProcessor.java
+++ b/core/devmode/src/main/java/io/quarkus/dev/RuntimeUpdatesProcessor.java
@@ -28,7 +28,6 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ConcurrentSkipListSet;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -80,16 +79,19 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext {
 
     public boolean doScan() throws IOException {
         final long startNanoseconds = System.nanoTime();
-        final ConcurrentMap<String, byte[]> changedClasses = scanForChangedClasses();
-        if (changedClasses == null)
-            return false;
 
-        DevModeMain.restartApp();
-        log.infof("Hot replace total time: %ss ", Timing.convertToBigDecimalSeconds(System.nanoTime() - startNanoseconds));
-        return true;
+        boolean classChanged = checkForChangedClasses();
+        boolean configFileChanged = checkForConfigFileChange();
+
+        if (classChanged || configFileChanged) {
+            DevModeMain.restartApp();
+            log.infof("Hot replace total time: %ss ", Timing.convertToBigDecimalSeconds(System.nanoTime() - startNanoseconds));
+            return true;
+        }
+        return false;
     }
 
-    ConcurrentMap<String, byte[]> scanForChangedClasses() throws IOException {
+    boolean checkForChangedClasses() throws IOException {
         final Set<File> changedSourceFiles;
 
         if (sourcesDir != null) {
@@ -103,35 +105,26 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext {
                         .collect(Collectors.toCollection(ConcurrentSkipListSet::new));
             }
         } else {
-            changedSourceFiles = Collections.EMPTY_SET;
+            changedSourceFiles = Collections.emptySet();
         }
         if (!changedSourceFiles.isEmpty()) {
-            log.info("Changes source files detected, recompiling " + changedSourceFiles);
-
+            log.info("Changed source files detected, recompiling " + changedSourceFiles);
             try {
                 compiler.compile(changedSourceFiles.stream()
                         .collect(groupingBy(this::getFileExtension, Collectors.toSet())));
             } catch (Exception e) {
                 DevModeMain.deploymentProblem = e;
-                return null;
+                return false;
             }
         }
-        final ConcurrentMap<String, byte[]> changedClasses;
         try (final Stream<Path> classesStream = Files.walk(classesDir)) {
-            changedClasses = classesStream
-                    .parallel()
-                    .filter(p -> p.toString().endsWith(".class"))
-                    .filter(p -> wasRecentlyModified(p))
-                    .collect(Collectors.toConcurrentMap(
-                            p -> pathToClassName(p),
-                            p -> CopyUtils.readFileContentNoIOExceptions(p)));
+            if (classesStream.parallel().anyMatch(p -> p.toString().endsWith(".class") && wasRecentlyModified(p))) {
+                // At least one class was recently modified
+                lastChange = System.currentTimeMillis();
+                return true;
+            }
         }
-        if (changedClasses.isEmpty() && !checkForConfigFileChange()) {
-            return null;
-        }
-
-        lastChange = System.currentTimeMillis();
-        return changedClasses;
+        return false;
     }
 
     private Optional<String> matchingHandledExtension(Path p) {
@@ -163,6 +156,7 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext {
                     Long existing = configFileTimestamps.get(i);
                     if (value > existing) {
                         ret = true;
+                        log.infof("Config file change detected: %s", config);
                         if (doCopy) {
                             Path target = classesDir.resolve(i);
                             byte[] data = CopyUtils.readFileContent(config);
@@ -201,12 +195,6 @@ public class RuntimeUpdatesProcessor implements HotReplacementContext {
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
-    }
-
-    private String pathToClassName(final Path path) {
-        String pathName = classesDir.relativize(path).toString();
-        String className = pathName.substring(0, pathName.length() - 6).replace('/', '.');
-        return className;
     }
 
     interface UpdateHandler {

